stdoutisatty: init at 1.0

bryango opened the pull request:
## Description of changes

Homepage: https://github.com/lilydjwg/stdoutisatty

It is surprisingly difficult to get consistent, colored outputs in the terminal, especially when commands are piped into a pager. In fact, there is a whole Arch wiki page on related topics:
- https://wiki.archlinux.org/title/Color_output_in_console#Reading_from_stdin

Among the solutions listed there, `stdoutisatty` is one of the simplest: it hijacks the [`isatty` function call](https://man.archlinux.org/man/isatty.3) used by most apps to determine whether `stdout` is an actual terminal, and always returns `1 (true)`. With this I can simply do:
```bash
stdoutisatty tree | less
cat flake.lock | stdoutisatty jq | less
...
```
... and enjoy my colorful outputs, without figuring out whatever `--color=always` type of flags these cli apps may or may not provide for colored outputs. Afaik, there is nothing like this in nixpkgs yet, so I would like to package and maintain it.

<!--
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

bryango committed:
stdoutisatty: init at 1.0

Homepage: https://github.com/lilydjwg/stdoutisatty

It is surprisingly difficult to get consistent, colored outputs
in the terminal, especially when commands are piped into a pager.
In fact, there is a whole Arch wiki page on related topics:

- https://wiki.archlinux.org/title/Color_output_in_console#Reading_from_stdin

Among the solutions listed there, `stdoutisatty` is one of the simplest:
it hijacks the `isatty` function call used by most apps to determine
whether `stdout` is an actual terminal, and always returns `1 (true)`.
With this I can simply do:

```bash
stdoutisatty tree | less
cat flake.lock | stdoutisatty jq | less
...
```

and enjoy my colorful outputs, without figuring out whatever
`--color=always` type of flags these cli apps may or may not provide
for colored outputs.

bryango committed:
stdoutisatty: add passthru.tests.ls-color

NickCao commented on pkgs/by-name/st/stdoutisatty/package.nix:
```diff
@@ -0,0 +1,70 @@
+
+  src = fetchFromGitHub {
+    owner = "lilydjwg";
+    repo = finalAttrs.pname;
```
This is discouraged: https://github.com/NixOS/nixpkgs/issues/310373

bryango replied:
Thank you for linking this interesting discussion! I didn't know. I have removed most superfluous use of `finalAttrs`. In particular, all fields of `src` are now spelled out explicitly. However, it does make me a bit uncomfortable to either:
- specify twice the version `"1.0"` which should clearly be identified, or
- use `rec` :laughing: 

So I opted for a slightly ugly fix: instead of defining `src.rev` with `version`, define `version` with `src.rev`:
```nix
  version = finalAttrs.src.rev;

  src = fetchFromGitHub {
    owner = "lilydjwg";
    repo = "stdoutisatty";
    rev = "1.0";
    hash = "sha256-NyVn9cxx0rY1ridNDTqe0pzcVhdLVaPCKT4hoQkQzRs=";
  };
```
Is it okay to do this? :laughing:

NickCao replied:
Actually it's fine wot use finalAttrs.version in rev. Just avoid pname is enough.

NickCao commented on pkgs/by-name/st/stdoutisatty/package.nix:
```diff
@@ -0,0 +1,70 @@
+    makeBinaryWrapper
+  ];
+
+  preFixup = ''
+    # necessary for dynamic loading of `lib/libstdoutisatty.so` at runtime
+    wrapProgram $out/bin/${finalAttrs.pname} \
+      --prefix LD_LIBRARY_PATH : "$out/lib"
+  '';
```
We should define `LIBDIR` instead: https://github.com/lilydjwg/stdoutisatty/blob/c77075065243f353a13c169941e14a6df7849038/stdoutisatty.c#L14, `LD_LIBRARY_PATH` is too big a hammer.

bryango replied:
Changed to use `cmakeFlags` in the newer version. In particular, since the upstream `CMakeLists.txt` specified `LIB_FILE`, we need to override `LIB_FILE` (instead of `LIBDIR`) for it to really work; see the logic around https://github.com/lilydjwg/stdoutisatty/blob/c77075065243f353a13c169941e14a6df7849038/stdoutisatty.c#L9-L16.

NickCao commented on pkgs/by-name/st/stdoutisatty/package.nix:
```diff
@@ -0,0 +1,70 @@
+        color_auto=$(ls -1 --color=auto)
+        color_always=$(ls -1 --color=always)
+
+        ${finalAttrs.finalPackage}/bin/${finalAttrs.meta.mainProgram} \
```
```suggestion
        ${lib.getExe finalAttrs.finalPackage} \
```

NickCao commented on pkgs/by-name/st/stdoutisatty/package.nix:
```diff
@@ -0,0 +1,70 @@
+  passthru = {
+    updateScript = nix-update-script { };
+    tests = {
+      ls-color = runCommand "${finalAttrs.pname}-ls-color" { } ''
```
Or we can use `tty -s` to simplify the test: https://stackoverflow.com/a/10022365

bryango replied:
This is interesting! Although I couldn't actually get `tty` to reliably determine whether `stdout` is a tty... For example, `faketty tty <&1 | less` prints `not a tty`, so I don't know how to proceed at the moment.

NickCao commented on pkgs/by-name/st/stdoutisatty/package.nix:
```diff
@@ -0,0 +1,70 @@
+    homepage = "https://github.com/lilydjwg/stdoutisatty";
+    license = lib.licenses.bsd2;
+    maintainers = with lib.maintainers; [ bryango ];
+    mainProgram = finalAttrs.pname;
```
Same problem as with `repo`

NickCao commented:
Also, we have faketty for the same purpose: https://github.com/dtolnay/faketty

bryango committed:
stdoutisatty: use cmakeFlags, not LD_LIBRARY_PATH

Co-authored-by: Nick Cao <nickcao@nichi.co>

bryango committed:
stdoutisatty: use `lib.getExe` in tests

Co-authored-by: Nick Cao <nickcao@nichi.co>

bryango committed:
stdoutisatty: avoid superfluous use of finalAttrs

The old version may lead to confusing `overrideAttrs` behaviors.
See discussions linked in the PR review comments.

Co-authored-by: Nick Cao <nickcao@nichi.co>

bryango commented:
> Also, we have faketty for the same purpose: [dtolnay/faketty](https://github.com/dtolnay/faketty)

Ah, this is awesome! Thank you! And it's Memory Safe (trademark) :laughing: 

So the existence of `faketty` makes `stdoutisatty` a lot less relevant... Nonetheless, I have addressed the issues with the packaging in case people are still interested in having it in nixpkgs.

NickCao commented on pkgs/by-name/st/stdoutisatty/package.nix:
```diff
@@ -0,0 +1,69 @@
+
+stdenv.mkDerivation (finalAttrs: {
+  pname = "stdoutisatty";
+  version = finalAttrs.src.rev;
```
```suggestion
  version = "1.0";
```

NickCao commented on pkgs/by-name/st/stdoutisatty/package.nix:
```diff
@@ -0,0 +1,69 @@
+  src = fetchFromGitHub {
+    owner = "lilydjwg";
+    repo = "stdoutisatty";
+    rev = "1.0";
```
```suggestion
    rev = finalAttrs.version;
```

NickCao commented on pkgs/by-name/st/stdoutisatty/package.nix:
```diff
@@ -0,0 +1,69 @@
+
+  cmakeFlags = [
+    # must specify the full path to `libstdoutisatty.so` in the nix store
+    "-DCMAKE_C_FLAGS=-DLIB_FILE='\"${placeholder "out"}/lib/libstdoutisatty.so\"'"
```
```suggestion
    (lib.cmakeFeature "CMAKE_C_FLAGS" "-DLIB_FILE='\"${placeholder "out"}/lib/libstdoutisatty.so\"'")
```

bryango replied:
Ah that's much better! Thank you!

bryango commented:
Fixed and squashed some commits (unsquashed: https://github.com/NixOS/nixpkgs/compare/master...7a5f7b92ebfa2a5e928fdc28a5b0b454942a6446)
